full-service/src/service/hardware_wallet.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2020-2021 MobileCoin Inc.
+
+//! Service for managing ledger materials and MobileCoin protocol objects.
+
+use std::convert::TryFrom;
+
+use ledger_mob::{DeviceHandle, Filters, LedgerHandle, LedgerProvider, Transport};
+
+use mc_common::logger::global_log;
+use mc_crypto_keys::RistrettoPublic;
+use strum::Display;
+
+use crate::service::models::tx_proposal::{InputTxo, TxProposal, UnsignedTxProposal};
+
+/// Errors for the Address Service.
+#[derive(Display, Debug)]
+#[allow(clippy::large_enum_variant)]
+pub enum HardwareWalletServiceError {
+    NoHardwareWalletsFound,
+    LedgerMob(ledger_mob::Error),
+    PresignedRingsNotSupported,
+    KeyImageNotFoundForSignedInput,
+    RingCT(mc_transaction_core::ring_ct::Error),
+    CryptoKeys(mc_crypto_keys::KeyError),
+}
+
+impl From<mc_transaction_core::ring_ct::Error> for HardwareWalletServiceError {
+    fn from(src: mc_transaction_core::ring_ct::Error) -> Self {
+        HardwareWalletServiceError::RingCT(src)
+    }
+}
+
+impl From<mc_crypto_keys::KeyError> for HardwareWalletServiceError {
+    fn from(src: mc_crypto_keys::KeyError) -> Self {
+        HardwareWalletServiceError::CryptoKeys(src)
+    }
+}
+
+impl From<ledger_mob::Error> for HardwareWalletServiceError {
+    fn from(src: ledger_mob::Error) -> Self {
+        HardwareWalletServiceError::LedgerMob(src)
+    }
+}
+
+async fn get_device_handle() -> Result<DeviceHandle<LedgerHandle>, HardwareWalletServiceError> {
+    let mut ledger_provider = LedgerProvider::init().await;
+    let devices = ledger_provider
+        .list(Filters::Hid)
+        .await
+        .map_err(ledger_mob::Error::from)?;
+
+    // Get the first device found, or error if none are found.
+    let device = devices
+        .first()
+        .ok_or(HardwareWalletServiceError::NoHardwareWalletsFound)?;
+
+    global_log::info!("Found devices: {:04x?}", devices);
+
+    let handle = ledger_provider
+        .connect(device.clone())
```

Comment:
Optional/Nit: No need to clone since we're not using `device` after this
```suggestion
        .connect(device)
```